src/util/storable_builder.rs

```diff
@@ -37,8 +37,8 @@ impl StorableBuilder {
 		Storable {
 			data: data_blob,
 			encryption_metadata: Some(EncryptionMetadata {
-				nonce: nonce.to_vec(),
-				tag: tag.to_vec(),
+				nonce: Vec::from(nonce),
+				tag: Vec::from(tag),
```

Comment:
These still result in heap allocations. A new array is created on the heap in `Vec` within a `Box` type and then the data is copied in because arrays implement `Copy`. Ultimately, the `nonce` and `tag` memory is already stack allocated.

Other than foregoing the use of the proto altogether or using an arena (if possible), the only way to avoid excessive heap allocations would be to allocate one `Storable` as a member of `StorableBuilder` and reuse its `Vec` fields by taking slices to use with `fill_bytes` and `encrypt_inplace`. Then you would only need to allocate one `String` for `cipher_format`, too. But the caller would need to also reuse `StorableBuilder`, and `build` couldn't return a `Storable` any more. Instead, the interface would be in terms of bytes and `Storable` would be an implementation detail.

Reply:
> Yeah but that introduces a state which mutates per request, making it non-thread safe. I think it will complicate everything.

You could use a `Storable` per thread, FWIW.

> But what you suggested gave me idea for another approach. Basically it will avoid Copy and array allocation but not vec allocation. We will create vec's in the first place and use references for fill_bytes and encrypt_inplace, similar to what you suggest.
> 
> Since these vec's are fixed size pre-allocated, i think it should be fine?

Eh, it removes the stack allocations, which are essentially free, and the copy, yeah.

> 
> (Overall i think it shouldn't be a big concern since it is around 28bytes and we allocate much more during just a single gossip msg forward)

The larger concern is not the allocation size but rather any resulting heap fragmentation. I guess your argument is that in comparison to other parts of the code (e.g., gossip) this is trivial or at least similar, IIUC. Most gossip messages are indeed larger but are allocated largely on the stack. Not entirely though as they do contain a `Vec` field here and there (e.g., features), so I'd buy your argument that perhaps this isn't too much to worry about from a frequency of small allocation perspective.

> 
> (From my limited understanding of glibc allocator, the optimization achieved by using a single Storable as field is easily achieved at allocator level i think. Repeated small allocations will reuse the same small-bins/fast-bins/tcache. It might not avoid alloc traffic but those operations are meant to be fast. As long as number of small allocations don't keep on growing it is fine.)

I'm not knowledgable enough here to say what would happen in practice. Essentially, we'll have three small, fixed-size heap allocations for the `Vec`s and `String`, which are free'ed whenever the `Storeable` is freed. Presumably the caller will call `encode` immediately to pass the value to the `KVStore`. Afterwards, the `Storeable` is freed upon drop and thus so are the small allocations.

The question is, will the allocator reuse those small pieces of memory the next time a `Storable` is created? And will there be much heap fragmentation in the interim?

@TheBlueMatt I'm indifferent on keeping the proto. Seems it's not a whole lot different from other places in the code where there may be comparable allocation patterns (e.g., gossip forwarding). But let me know if I'm missing anything.

Reply:
> No they are heap allocated, MessageBuf is just an abstraction over Vec [link](https://github.com/lightningdevkit/rust-lightning/pull/2708/commits/969085bf1e77ebe5b4e7cb0523311e9905fa20f3)

FWIW, I was referring to the gossip messages _after_ they have been deserialized from a `MessageBuf`. Those structs' fields are mostly (but not entirely) primitives or other structs that don't require heap allocation.